comment/models/reactions.py

```diff
@@ -0,0 +1,199 @@
+            self.save()
+
+    def _decrease_dislikes(self):
+        """Decrease dislikes and save the model"""
+        if self.dislikes > 0:
+            self.dislikes = models.F('dislikes') - 1
+            self.save()
+
+    def _increase_reaction_count(self, reaction):
+        """
+        Increase reaction count(likes/dislikes)
+
+        Args:
+            reaction ([int]): The integral value that matches to the reaction value in
+                the database
+            
+        Returns:
+            None
+        """
+        if reaction == ReactionInstance.ReactionType.LIKE.value:
+            self._increase_likes()
+        else:
+            self._increase_dislikes()
+
+    def _decrease_reaction_count(self, reaction):
+        """
+        Decrease reaction count(likes/dislikes)
+
+        Args:
+            reaction ([int]): The integral value that matches to the reaction value in
+                the database
+
+        Returns:
+            None
+        """
+        if reaction == ReactionInstance.ReactionType.LIKE.value:
+            self._decrease_likes()
+        else:
+            self._decrease_dislikes()
+
+
+class ReactionInstance(models.Model):
+    
+    @unique
+    class ReactionType(IntEnum):
+        LIKE = 1
+        DISLIKE = 2
+
+    reaction_choices = [(r.value, r.name) for r in ReactionType] 
+    
+    reaction = models.ForeignKey(Reaction, related_name='reactions', on_delete=models.CASCADE)
+    user = models.ForeignKey(get_user_model(), related_name='users', on_delete=models.CASCADE)
+    reaction_type = models.SmallIntegerField(choices=reaction_choices)
+    date_reacted = models.DateTimeField(auto_now=timezone.now())
+
+    class Meta:
+        unique_together = ['user', 'reaction']
+
+    @classmethod
+    def clean_reaction_type(cls, reaction_type):
+        """
+        Check if the reaction is a valid one or not
+
+        Args:
+            reaction_type (str): The reaction to be saved
+
+        Returns:
+            int: The integral value that matches to the reaction value in
+                the database
+        """
+        reaction = getattr(cls.ReactionType, reaction_type.upper(), None)
+        if not reaction:
+            return ValidationError(
+                _('%(reaction)s is an invalid reaction'), 
+                code='invalid',
+                params={'reaction':reaction}
+                )
+        return reaction.value
+
+    def save(self, *args, **kwargs):
+        """
+        Increase reaction count in the reaction model after saving an instance
+
+        Args:
+            reaction_type (int): The integral value that matches to the reaction value in
+                the database
+        """
+        super().save(*args, **kwargs)
+        reaction = self.reaction
+        reaction_type = self.reaction_type
+        reaction._increase_reaction_count(reaction_type)
+
+    @classmethod
+    def _delete_and_create_new_instance(cls, instance, user, reaction_type):
```

Comment:
What the purpose of this method?
Can we adopt the following workflow:
* Whenever a new comment is created, a new Reaction entry will be created as well (use signals).
* When the user reacts on a comment for the first time, the ReactionInstance entry will be created. This entry will live as long as both the user and the comment are there, user or comment is deleted, the entry will be deleted.
* Further reactions from the user:

    - `like` => in Reaction model `increate_likes` called and in ReactionInstance `reaction_type` is set to 1
    - `dislike` same above with corresponding values.
    - `unlike` or `undislike` When the user press again on the liked or disliked icon the corresponding reaction should be decreased and the `reaction_type` in the ReactionInstance should be set to `(0, NEUTRAL)`

Reply:
This function deletes a previous instance of `ReactInstance` that the user had and creates a new one if required.
Use cases
- The user has already liked a comment and they again make a request to like the comment, it will only delete the previous instance. 
-  The user has already liked a comment and they make a request to dislike the comment, tit will delete the previous instance and create a new instance for dislike.




>     * `unlike` or `undislike` When the user press again on the liked or disliked icon the corresponding reaction should be decreased and the `reaction_type` in the ReactionInstance should be set to `(0, NEUTRAL)`

Why do we need a third reaction? We can just delete the previous instance. The user could have mistakenly clicked on the like icon. In case you want it just as a default value for the database, is it necessary in our case?

Reply:
The instance is actually a db entry, my suggestion is to avoid deleting it. This will give us idea of the user activity. NUTRAL type will show us that the use has reacted with a comment.

Reply:
Yes I do get your point but the user could also mistakenly click on one of the reaction. It would not always be a correct measure of their reaction. It could be useful if the reactions weren't mutually exclusive.